oss-licenses-plugin/src/main/groovy/com/google/android/gms/oss/licenses/plugin/OssLicensesPlugin.groovy

```diff
@@ -16,75 +16,76 @@
 import com.android.build.gradle.api.BaseVariant
 import org.gradle.api.Plugin
 import org.gradle.api.Project
+import org.slf4j.LoggerFactory
 
 class OssLicensesPlugin implements Plugin<Project> {
+
+    private static final logger = LoggerFactory.getLogger(DependencyTask.class)
+
     void apply(Project project) {
-        def getDependencies = project.tasks.create("getDependencies",
-                DependencyTask)
-        def dependencyOutput = new File(project.buildDir,
-            "generated/third_party_licenses")
-        def generatedJson = new File(dependencyOutput, "dependencies.json")
-        getDependencies.setProject(project)
-        getDependencies.outputDir = dependencyOutput
-        getDependencies.outputFile = generatedJson
+        def variantTolicenseTaskMap = new HashMap<String, LicensesTask>()
+        project.androidComponents {
+            onVariants(selector().all(), { variant ->
+                def baseDir = new File(project.buildDir,
+                        "generated/third_party_licenses/${variant.name}")
+                def dependenciesJson = new File(baseDir, "dependencies.json")
 
-        def resourceOutput = new File(dependencyOutput, "/res")
-        def outputDir = new File(resourceOutput, "/raw")
-        def licensesFile = new File(outputDir, "third_party_licenses")
-        def licensesMetadataFile = new File(outputDir,
-                "third_party_license_metadata")
-        def licenseTask = project.tasks.create("generateLicenses", LicensesTask)
+                def dependencyTask = project.tasks.register(
+                        "${variant.name}OssDependencyTask",
+                        DependencyTask.class) {
+                    it.outputDir = baseDir
+                    it.outputFile = dependenciesJson
+                    it.libraryDependenciesReport.set(variant.artifacts.get(SingleArtifact.METADATA_LIBRARY_DEPENDENCIES_REPORT.INSTANCE))
+                }.get()
+                logger.debug("Created task ${dependencyTask.name}")
 
-        licenseTask.dependenciesJson = generatedJson
-        licenseTask.outputDir = outputDir
-        licenseTask.licenses = licensesFile
-        licenseTask.licensesMetadata = licensesMetadataFile
+                def resourceBaseDir = new File(baseDir, "/res")
+                def rawResourceDir = new File(resourceBaseDir, "/raw")
+                def licensesFile = new File(rawResourceDir, "third_party_licenses")
+                def licensesMetadataFile = new File(rawResourceDir,
+                        "third_party_license_metadata")
 
-        licenseTask.inputs.file(generatedJson)
-        licenseTask.outputs.dir(outputDir)
-        licenseTask.outputs.files(licensesFile, licensesMetadataFile)
+                def licenseTask = project.tasks.register(
+                        "${variant.name}OssLicensesTask",
+                        LicensesTask.class) {
+                    it.dependenciesJson = dependenciesJson
```

Comment:
avoid implicit dependency like this and instead declare `it.dependencies` as a RegularFileProperty and do : 

`it.dependenciesJson.set(dependencyTask.flatMap { task -> task.dependenciesJson }`

then you can remove the line 60 as the dependency is explicit with the statement above.

Reply:
Done, except dependencyTask wasn't a collection so I didn't flatMap.